### PR TITLE
fixes typing errors in auto_restart.py

### PR DESCRIPTION
**Pull request by donlapark**
## What does this PR do?

Fixes typing errors in `pytorch_lightning/utilities/auto_restart.py` as a part of #13445.

Well, this was quite a tough one, as the type of `state_dict` is not consistent throughout the file. Sometimes its type is `Dict[str, Any]`, the other times, it is `Dict[int, Any]`.

Also, the `state` attribute in `MergedIteratorState` can have two types: `Dict[int, IteratorState]` or `Dict[str, Dict[int, IteratorState]]`, depending on whether the dataset is map-based or iterable. I have applied a band-aid fix which initially sets the type of `state` to `Dict` and then conditionally sets it to a more specific type in `update` method later.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/utilities/auto_restart.py [hunk 1]
```diff
@@ -268,17 +296,18 @@ def __init__(self, dataset: IterableDataset) -> None:
...
     @property
     def sampler(self) -> Sampler:
         return self.dataset.sampler
 
     def state_dict(self) -> Dict[str, Any]:
+        assert self.samplers is not None
         return {k: v.state_dict() for k, v in self.samplers.items()}
 
-    def load_state_dict(self, state_dict: Dict[int, Any]) -> None:
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
```
Comment (awaelchli) [inline src/pytorch_lightning/utilities/auto_restart.py:310]: If this is getting changed, then why do we introduce the IntStateful type? I'm a little confused about this.

File: src/pytorch_lightning/utilities/auto_restart.py [hunk 2]
```diff
@@ -24,14 +25,35 @@
...
 from pytorch_lightning.utilities.enums import _FaultTolerantMode, AutoRestartBatchKeys
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
 from pytorch_lightning.utilities.seed import _collect_rng_states, _set_rng_states
-from pytorch_lightning.utilities.types import _Stateful
+from pytorch_lightning.utilities.types import _IntStateful, _Stateful
+
+if TYPE_CHECKING:
+    _BaseLoaderIter = _BaseDataLoaderIter
+else:
+    _BaseLoaderIter = object
```
[the remark this answers is not in the document]
Comment (donlapark) [reply src/pytorch_lightning/utilities/auto_restart.py:41]: Hmm..I already forgot the purpose of this. Let's just remove it.

```suggestion
```

File: src/pytorch_lightning/utilities/auto_restart.py (hunk 1 above)
Comment (donlapark) [reply src/pytorch_lightning/utilities/auto_restart.py:305]: The `_IntStateful` was added because some input of `load_state_dict` is `Dict[int, Any]`. In particular,
https://github.com/Lightning-AI/lightning/blob/f3ef4f8b4d5231442b0305997da718a2684040e0/src/pytorch_lightning/utilities/auto_restart.py#L604-L605

The output of `_rotate_worker_indices` is expected to be `Dict[int, Any]`.
https://github.com/Lightning-AI/lightning/blob/f3ef4f8b4d5231442b0305997da718a2684040e0/src/pytorch_lightning/utilities/auto_restart.py#L626

**Review by rohitgr7 — APPROVED** [review]

**Comment by rohitgr7** [comment]
@carmocca mind review as code-owner?

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by Borda — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: src/pytorch_lightning/utilities/types.py [hunk 3]
```diff
@@ -101,6 +101,18 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
         ...
 
 
+@runtime_checkable
+class _IntStateful(Protocol):
```
Comment (carmocca) [inline src/pytorch_lightning/utilities/types.py:105]: I think this could be done with Generics where `_Stateful` takes a `Generic[T]` which here is `int`

Comment (donlapark) [reply src/pytorch_lightning/utilities/types.py:105]: Alright, let us try modifying `_Stateful`.

Comment (donlapark) [reply src/pytorch_lightning/utilities/types.py]: The following modification passes a `mypy` check.

https://github.com/Lightning-AI/lightning/blob/c11d3e03ea8923f25d098478328ab60da6e9a325/src/pytorch_lightning/utilities/types.py#L96-L102

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/utilities/types.py (hunk 3 above)
Comment (carmocca) [reply src/pytorch_lightning/utilities/types.py]: Thanks! Pushed https://github.com/Lightning-AI/lightning/pull/13904/commits/71d81da6d870b16b526297ee05d658affd44fe7c to finish it
